cassandra/io/eventletreactor.py

```diff
@@ -89,6 +101,34 @@ def __init__(self, *args, **kwargs):
         self._write_watcher = eventlet.spawn(lambda: self.handle_write())
         self._send_options_message()
 
+    def _wrap_socket_from_context(self):
+        if isinstance(self.ssl_context, eventlet.green.ssl.GreenSSLContext):
+            super(EventletConnection, self)._wrap_socket_from_context()
+        else:
+            _check_pyopenssl()
+            self._socket = SSL.Connection(self.ssl_context, self._socket)
+            self._socket.set_connect_state()
+            if self.ssl_options and 'server_hostname' in self.ssl_options:
+                # This is necessary for SNI
+                self._socket.set_tlsext_host_name(self.ssl_options['server_hostname'].encode('ascii'))
+
+    def _initiate_connection(self, sockaddr):
+        if isinstance(self.ssl_context, eventlet.green.ssl.GreenSSLContext):
+            super(EventletConnection, self)._initiate_connection(sockaddr)
+        else:
+            self._socket.connect(sockaddr)
+            if self.ssl_context or self.ssl_options:
+                self._socket.do_handshake()
+
+    def _match_hostname(self):
+        if isinstance(self.ssl_context, eventlet.green.ssl.GreenSSLContext) or not self.ssl_context:
+            super(EventletConnection, self)._match_hostname()
+        else:
+            cert_name = self._socket.get_peer_certificate().get_subject().commonName
+            if cert_name != self.endpoint.address:
+                raise Exception("Hostname verification failed! Certificate name '{}' "
+                                "doesn't endpoint '{}'".format(cert_name, self.endpoint.address))
+
```

Comment:
Yes. test_ssl.py has some tests using `ssl_options` and other tests using `ssl_context`. The `ssl_options` tests will take the `if` branch while the `ssl_context` tests will take the `else` branch.

Reply:
I was refering to different possible ssl context types. f I follow correctly, `self._wrap_socket_from_context()` is called only when there is a ssl_context. So there now 3 possibilities to test, right?
1. just ssl_options (legacy)
2. isinstance(self.ssl_context, eventlet.green.ssl.GreenSSLContext) == True
3. isinstance(self.ssl_context, eventlet.green.ssl.GreenSSLContext) == False

Reply:
Ah, right. 1 and 2 are the same. That first if condition you pointed out should never happen. I'll remove it.

How about I add something like a `self.uses_legacy_ssl_options = isinstance(self.ssl_context, eventlet.green.ssl.GreenSSLContext)` in the `__init__` and then check that instead. It would make the reading of those methods a little clearer.

Reply:
+1 for that. Thanks